# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmTypeDef.h
# Commit: 72bf9c871ded
# Full Hash: 72bf9c871ded006f56f09b24d0723365516445bb
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 5
# Description:
#   Bug 1692065 - wasm: Add tagging for 31-bit integers and i31ref. r=yury
#   
#   This commit adds a tag 0x1 for 31-bit integers in AnyRef. Any JS value
#   that is losslessly representable (without coercion) in this range is
#   stored without boxing. This is chosen to line up exactly with how the
# ==============================================================================

diff -r 28feda14fdec -r 72bf9c871ded js/src/wasm/WasmTypeDef.h
--- a/js/src/wasm/WasmTypeDef.h	Fri Aug 04 14:06:50 2023 +0000
+++ b/js/src/wasm/WasmTypeDef.h	Fri Aug 04 14:06:50 2023 +0000
@@ -1293,6 +1293,7 @@
       return RefTypeHierarchy::Extern;
     case RefType::Any:
     case RefType::None:
+    case RefType::I31:
     case RefType::Eq:
     case RefType::Struct:
     case RefType::Array:
@@ -1356,6 +1357,11 @@
     return true;
   }
 
+  // i31ref is a subtype of eqref
+  if (subType.isI31() && (superType.isAny() || superType.isEq())) {
+    return true;
+  }
+
   // structref/arrayref are subtypes of eqref and anyref
   if ((subType.isStruct() || subType.isArray()) &&
       (superType.isAny() || superType.isEq())) {